File: Dynamic Programming/11066번 - 파일 합치기/11066.cpp
#include <iostream>
#include <vector>

using namespace std;

#define INF 1e9
#define LIM 501

int novel;
vector<int> pages;
int sum[LIM] = { 0, };
int dp[LIM][LIM] = { 0, };

void input() {
    cin >> novel;
    pages.clear();
    pages.resize(novel + 1);
    
    for (int i = 1; i <= novel; i++) {
        cin >> pages[i];
    }
}

void solve() {
    for(int i = 1; i <= novel; i++) {
        sum[i] = sum[i - 1] + pages[i];
    }
    
    // 임의의 k만큼의 구간
    for (int k = 1; k < novel; k++) {
        // 시작 위치
        for (int st = 1; st + k <= novel; st++) {
            int end = st + k;
            dp[st][end] = INF;
            
            int sum_page = sum[end] - sum[st - 1];
            // 중간 부분으로 나눠서 최솟값 계산 
            for (int mid = st; mid < end; mid++) {
                dp[st][end] = min(dp[st][end],
                                  dp[st][mid] + dp[mid + 1][end] + sum_page);
            }
        }
    }
    
    cout << dp[1][novel] << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int tc = 0;
    cin >> tc;
    while(tc--) {
        input();
        solve();
    }
    
    return 0;
}

File: Dynamic Programming/11066번 - 파일 합치기/11066.java
import java.io.*;
import java.util.*;

public class Main {
    static final int INF = (int) 1e9;
    static final int LIM = 501;

    static int novel;
    static int[] pages = new int[LIM];
    static int[] sum = new int[LIM];
    static int[][] dp = new int[LIM][LIM];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine());

        while (tc-- > 0) {
            novel = Integer.parseInt(br.readLine());
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= novel; i++) {
                pages[i] = Integer.parseInt(st.nextToken());
            }

            // dp, sum 초기화
            for (int i = 0; i <= novel; i++) {
                sum[i] = 0;
                Arrays.fill(dp[i], 0);
            }

            solve();
        }
    }
    
    static void solve() {
        for (int i = 1; i <= novel; i++) {
            sum[i] = sum[i - 1] + pages[i];
        }

        for (int k = 1; k < novel; k++) { // k - 1 구간
            for (int st = 1; st + k <= novel; st++) { // 시작점
                int end = st + k; // 끝점
                dp[st][end] = INF;

                int sum_page = sum[end] - sum[st - 1];
                for (int mid = st; mid < end; mid++) {
                    dp[st][end] = Math.min(dp[st][end],
                                           dp[st][mid] + dp[mid + 1][end] + sum_page);
                }
            }
        }

        System.out.println(dp[1][novel]);
    }
}
